core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala

```diff
@@ -56,6 +57,28 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
   //      SQLPlanParser.
   var unsupportedSQLPlan: ArrayBuffer[UnsupportedSQLPlan] = ArrayBuffer[UnsupportedSQLPlan]()
   var allSQLMetrics: ArrayBuffer[SQLMetricInfoCase] = ArrayBuffer[SQLMetricInfoCase]()
+  // A map between stage ID and the set of node names
+  val stageToNodeNames: HashMap[Long, Seq[String]] = HashMap.empty[Long, Seq[String]]
+  // A map between stage ID and diagnostic metrics values
+  val stageToDiagnosticMetrics: HashMap[Long, HashMap[String, StatisticsMetrics]] =
+    HashMap.empty[Long, HashMap[String, StatisticsMetrics]]
```

Comment:
Refactored implementation based on our offline discussion.
The high-level approach is that we create one `AccumProfileResults` object and reuse its reference in the mapping `stageToDiagnosticMetrics` to avoid allocating extra memory from creating new objects.
We have not find a good way to optimize `stageToNodeNames` yet.